packages/arb-token-bridge-ui/src/hooks/useCompleteMultiChainTransactions.ts

```diff
@@ -0,0 +1,323 @@
+        l1Provider: parentChainProvider,
+        l2Provider: chainProvider
+      })
+    }
+  }
+
+  if (withdrawal) {
+    return transformWithdrawal(withdrawal)
+  }
+}
+
+function getProvider(chainId: ChainId) {
+  const rpcUrl =
+    rpcURLs[chainId] ?? getWagmiChain(chainId).rpcUrls.default.http[0]
+  return new StaticJsonRpcProvider(rpcUrl)
+}
+
+/**
+ * Fetches transaction history only, and only for a specific direction.
+ * The query could be e.g. deposits or withdrawals.
+ */
+const useTransactionListByDirection = (
+  direction: 'deposits' | 'withdrawals'
+) => {
+  const [transactions, setTransactions] = useState<DepositOrWithdrawal[][]>([])
+  const [page, setPage] = useState(0)
+  const [loading, setLoading] = useState(true)
+
+  const { address } = useAccount()
+
+  const shouldFetchNextPageForChainPair = useCallback(
+    (chainPairIndex: number) => {
+      if (page === 0) {
+        return true
+      }
+
+      const txCountForChainPair = transactions[chainPairIndex]?.length ?? 0
+      // fetch next page for a chain pair if all fetched pages are full
+      // we check for >= in case some txs were included by initiating a transfer
+      return txCountForChainPair / page >= PAGE_SIZE
+    },
+    [page, transactions]
+  )
+
+  const { data, error } = useSWRImmutable(
+    address ? ['tx_list', direction, address, page] : null,
+    ([, _direction, _address, _page]) => {
+      return Promise.all(
+        multiChainFetchList.map((chainPair, chainPairIndex) => {
+          if (!shouldFetchNextPageForChainPair(chainPairIndex)) {
+            return []
+          }
+
+          const params = {
+            sender: _address,
+            receiver: _address,
+            l1Provider: getProvider(chainPair.parentChain),
+            l2Provider: getProvider(chainPair.chain),
+            pageNumber: _page,
+            pageSize: PAGE_SIZE
+          }
+
+          return _direction === 'deposits'
+            ? fetchDepositList(params)
+            : fetchWithdrawalList(params)
+        })
+      )
+    }
+  )
+
+  useEffect(() => {
+    if (!data) {
+      return
+    }
+
+    const fetchedSomeData = data.some(item => item.length > 0)
+
+    if (fetchedSomeData) {
+      // include the new data with the previously fetched data
+      // the data is grouped by chain pairs
+      setTransactions(prevTransactions => {
+        return data.map((transactionsForChainPair, chainPairIndex) => [
+          ...(prevTransactions[chainPairIndex] ?? []),
+          ...(transactionsForChainPair ?? [])
+        ])
+      })
+
+      setPage(prevPage => prevPage + 1)
+    } else {
+      setLoading(false)
+    }
+  }, [data])
+
+  return {
+    data: transactions.flat(),
+    loading,
+    error
+  }
+}
+
+/**
+ * Fetches transaction history only, without mapping additional info that would require a lot of RPC calls.
+ * The data is collected for multiple chain pairs, and sorted by date.
+ */
+const useMultiChainTransactionList = () => {
```

Comment:
I feel like this should belong to the `useTransactionListByDirection` so it gives us merged deposits & withdrawals, sorted and everything. We can remove the `direction` prop.

Reply:
Then we end up with 2 hooks instead of 3.

Reply:
done